Bowling.java

```diff
@@ -116,28 +120,50 @@ private Frame getCurrentFrame() {
 		return currentFrame;
 	}
 	
-	private Optional<Frame> previousFrame() {
-		if (allFrames.size() > 1) {
-			int nextToLastIndex = allFrames.size() - 2;
+	private Optional<Frame> getFrameBefore(Frame frame) {
+		int currentIndex = allFrames.indexOf(frame);
+		int lastIndex = allFrames.size() - 1;
+
+		if (allFrames.size() > 1 && currentIndex == lastIndex) {
+			int nextToLastIndex = allFrames.indexOf(frame) - 1;
 			return Optional.of(allFrames.get(nextToLastIndex));
+		} else if (allFrames.size() > 2) {
+			int thirdToLastIndex = allFrames.size() - 3;
+			return Optional.of(allFrames.get(thirdToLastIndex));
 		}
```

Comment:
What about this thinking?  You want to know if the currentIndex has one before it - then you just need to know that currentIndex > 0.  So I think this section could be changed to the following:

```
private Optional<Frame> getFrameBefore(Frame frame) {
    int currentIndex = allFrames.indexOf(frame);

    if (currentIndex > 0) {
        return Optional.of(allFrames.get(currentIndex - 1);
    }

    return Optional.empty();
}
```

Now that if check doesn't speak very well, so if you really wanted to you could create a helper method like `currentFrameIsNotFirstFrame(frame)` - maybe - not sure I'm thrilled with that.